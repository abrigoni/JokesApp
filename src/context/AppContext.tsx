import React, {createContext, FC, useEffect, useState} from 'react';
import { JokeSchema, JOKE_SCHEMA_NAME } from '../realm/schemas';
import {Joke} from '../types/Joke';
import Realm from 'realm';

interface ContextProps {
  savedJokes: Joke[];
  saveJoke: (joke: Joke) => void;
  removeJoke: (id: string) => void;
}

export const AppContext = createContext<ContextProps>({
  savedJokes: [],
  saveJoke: (_joke: Joke) => {},
  removeJoke: (_id: string) => {},
});

const AppContextProvider: FC = ({children}) => {
  const [realm, setRealm] = useState<Realm>();
  const [savedJokes, setSavedJokes] = useState<Joke[]>([]);

  // initial load on realm connection + savedJokes from DB
  useEffect(() => {
    const openRealm = async () => {
      const r = await Realm.open({schema: [JokeSchema]});
      if (r != null && !r.isClosed) {
        setRealm(r);
        const result = r.objects<Joke>(JOKE_SCHEMA_NAME).map((value) => {
          return value.toJSON() as Joke;
        });
        setSavedJokes(result);
      }
    };
    openRealm();
    return () => {
      if (!!realm && !realm.isClosed) {
        realm.close();
      }
    };
  }, []);

  const realm_saveJoke = (joke: Joke) => {
    if (!!realm && !realm.isClosed) {
      realm.write(() => {
        realm.create(JOKE_SCHEMA_NAME, joke);
      });
    }
  };

  const realm_deleteJoke = (id: string) => {
    if (!!realm && !realm.isClosed) {
      realm.write(() => {
        const joke = realm.objectForPrimaryKey<Joke>(JOKE_SCHEMA_NAME, id);
        realm.delete(joke); 
      });
    }
  }

  const saveJoke = (joke: Joke) => {
    if (!savedJokes.includes(joke)) {
      setSavedJokes((jokes: Joke[]) => [...jokes, joke]);
      realm_saveJoke(joke);
    }
  };

  const removeJoke = (id: string) => {
    setSavedJokes((jokes: Joke[]) =>
      jokes.filter((joke: Joke) => joke.id !== id),
    );
    realm_deleteJoke(id);
  };

  return (
    <AppContext.Provider
      value={{
        savedJokes,
        removeJoke,
        saveJoke,
      }}>
      {children}
    </AppContext.Provider>
  );
};

export default AppContextProvider;
