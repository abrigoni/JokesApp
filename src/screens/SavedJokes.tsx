import React, { FC, useContext, useLayoutEffect } from 'react'
import { Text, View, StyleSheet, Dimensions, FlatList } from 'react-native'
import IconButton from '../components/IconButton';
import JokeCard from '../components/JokeCard';
import { OpenSansText } from '../components/Typography';
import { AppContext } from '../context/AppContext';
import { Colors } from '../utils/colors';

export const SAVED_JOKES_ROUTE = "SavedJokes";

const { width, height } = Dimensions.get('window');

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    paddingVertical: 24,
    paddingHorizontal: width * 0.05,
  },
  card: {
    height: height * 0.25,
    margin: 10,
  },
  title: {
    fontSize: 12,
  },
  author: {
    fontSize: 16,
  },
});

interface Props {
  navigation: any; // todo
}

const SavedJokes: FC<Props> = ({navigation}) => {
  const { savedJokes, removeJoke } = useContext(AppContext);
  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => <OpenSansText size="Body" variant="Bold">{savedJokes.length}</OpenSansText>,
    });
  }, [navigation, savedJokes]);

  const handleDelete = (id: string) => {
    removeJoke(id);
  };

  return (
    <View style={styles.screen}>
      <FlatList
        data={savedJokes}
        numColumns={2}
        renderItem={({ item }) => (
            <JokeCard
              content={item.joke}
              backgroundColor={Colors.orange}
              style={styles.card}
              titleStyle={styles.title}
              authorStyle={styles.author}
              BottomIcon={<IconButton iconProps={{name: 'trash', size: 20,}} onPress={handleDelete.bind(this, item.id)} />}
            />
          )
        }
      />
    </View>
  )
}

export default SavedJokes