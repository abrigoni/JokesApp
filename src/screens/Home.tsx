import React, { FC, useContext, useLayoutEffect, useState } from 'react'
import {
  SafeAreaView,
  StyleSheet,
  Text,
  View,
  Dimensions,
  Pressable,
} from 'react-native';
import { Colors } from '../utils/colors';
import JokeCard from '../components/JokeCard';
import Button from '../components/Button';
import Carousel from '../components/Carousel';
import { AppContext } from '../context/AppContext';
import { OpenSansText } from '../components/Typography';
import { SAVED_JOKES_ROUTE } from './SavedJokes';
import Icon from 'react-native-vector-icons/FontAwesome';
import { Joke } from '../types/Joke';


const { width } = Dimensions.get('window');

const DUMMY_JOKES: Joke[] = [
  {
    id: "j1",
    joke: "Anyway. You're the designer and you know that to do best.",
  },
    {
    id: "j2",
    joke: "Mmmh. Do you think it will be easy to read? Let's make it bigger ðŸ˜‚",
  },
    {
    id: "j3",
    joke: "There's something that doesn't work. But I'm not sure what.",
  },
    {
    id: "j4",
    joke: "Anyway. You're the designer and you know that to do best.",
  },
];

const styles = StyleSheet.create({
  screen: {
    flex: 1,
  },
  screenContent: {
    flex: 1,
    paddingBottom: 24,
    paddingHorizontal: width * 0.05,
  },
  title: {
    fontFamily: 'OpenSans-Bold',
    fontSize: 34,
    color: 'black',
    textAlign: 'center',
  },
  cardContainer: {
    marginVertical: 32,
    flex: 1,
  },
});

const JOKE_CARD_COLORS: string[] = [
  Colors.purple,
  Colors.orange,
  Colors.red,
  Colors.green,
];


const renderJokeItem = (scrollX: any, {item}: {item: Joke}) => {
  const backgroundColor = scrollX.interpolate({
    inputRange: JOKE_CARD_COLORS.map((_, i) => i * width),
    outputRange: JOKE_CARD_COLORS.map((bg) => bg),
  });
  return <JokeCard content={item.joke} backgroundColor={backgroundColor}/>;
};

export const HOME_ROUTE = 'Home';

interface Props {
  navigation: any;
};

const Home: FC<Props> = ({navigation}) => {
  const {saveJoke, savedJokes} = useContext(AppContext);
  const [currentIndex, setCurrentIndex] = useState(0);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Pressable style={{flexDirection: 'row'}} onPress={() => navigation.navigate(SAVED_JOKES_ROUTE)}>
          <OpenSansText size="Body" variant="Bold">Saved: {savedJokes.length} </OpenSansText>
          <Icon name="heart" color="#e74c3c" size={24} />
        </Pressable>
      ),
    });
  }, [navigation, savedJokes]);

  const handleChangeIndex = (index: number) => {
    setCurrentIndex(index);
  };

  const handleSave = () => {
    saveJoke(DUMMY_JOKES[currentIndex]);
  };
  return (
    <SafeAreaView style={styles.screen}>
      <View style={styles.screenContent}>
        <View>
          <Text style={styles.title}>Things you can say</Text>
          <Text style={styles.title}>to annoy designers.</Text>
        </View>
        <View style={styles.cardContainer}>
          <Carousel items={DUMMY_JOKES} renderItem={renderJokeItem} updateIndex={handleChangeIndex} />
        </View>
        <Button title="Save" onPress={handleSave} />
      </View>
    </SafeAreaView>
  )
}

export default Home