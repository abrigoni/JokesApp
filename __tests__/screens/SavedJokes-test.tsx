import React from 'react';
import Home from '../../src/screens/Home';
import {MockedProvider} from '@apollo/client/testing';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { fireEvent, render } from "@testing-library/react-native";
import { AppContext } from "../../src/context/AppContext";
import SavedJokes from "../../src/screens/SavedJokes";

const navigation = {
  setOptions: jest.fn(),
  navigate: jest.fn(),
};

const mocks: any[] = [];

const removeJoke = jest.fn();

describe('<SavedJokes />', () => {
  test('renders correctly', () => {
    const component = (
      <MockedProvider mocks={mocks}>
        <AppContext.Provider value={{
          saveJoke: jest.fn(),
          removeJoke: jest.fn(),
          savedJokes: [
            {
              id: '1',
              joke: 'Test'
            },
            {
              id: '2',
              joke: 'Test 2'
            }]
        }}>
          <SavedJokes navigation={navigation} />
        </AppContext.Provider>
      </MockedProvider>
    );
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
  test('Click remove in card calls context removeJoke', async () => {
    const component = (
      <MockedProvider mocks={mocks}>
        <AppContext.Provider value={{
          saveJoke: jest.fn(),
          removeJoke,
          savedJokes: [
            {
              id: '1',
              joke: 'Test'
            }]
        }}>
          <SavedJokes navigation={navigation} />
        </AppContext.Provider>
      </MockedProvider>
    );
    const {getByText} = render(component);
    const toClick = await getByText('ï‡¸');
    fireEvent(toClick, 'press');
    expect(removeJoke).toHaveBeenCalledTimes(1);
    expect(removeJoke).toHaveBeenCalledWith('1');
  });
});
