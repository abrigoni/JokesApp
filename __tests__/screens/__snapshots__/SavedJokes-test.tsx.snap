// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SavedJokes /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingHorizontal": 37.5,
      "paddingVertical": 24,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "1",
          "joke": "Test",
        },
        Object {
          "id": "2",
          "joke": "Test 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f39c12",
                "borderRadius": 8,
                "flex": 1,
                "height": 333.5,
                "justifyContent": "space-between",
                "margin": 10,
                "marginHorizontal": 20,
                "paddingBottom": 30,
                "paddingHorizontal": 20,
                "paddingTop": 80,
                "width": 600,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#ecf0f1",
                    "fontFamily": "OpenSans-Bold",
                    "fontSize": 24,
                    "textAlign": "center",
                  },
                  Object {
                    "fontSize": 12,
                  },
                ]
              }
            >
              Test
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#ecf0f1",
                      "fontFamily": "IndieFlower",
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "marginRight": 8,
                    },
                  ]
                }
              >
                @designhumor
              </Text>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 20,
                      },
                      undefined,
                      Object {
                        "fontFamily": "FontAwesome",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f39c12",
                "borderRadius": 8,
                "flex": 1,
                "height": 333.5,
                "justifyContent": "space-between",
                "margin": 10,
                "marginHorizontal": 20,
                "paddingBottom": 30,
                "paddingHorizontal": 20,
                "paddingTop": 80,
                "width": 600,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#ecf0f1",
                    "fontFamily": "OpenSans-Bold",
                    "fontSize": 24,
                    "textAlign": "center",
                  },
                  Object {
                    "fontSize": 12,
                  },
                ]
              }
            >
              Test 2
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#ecf0f1",
                      "fontFamily": "IndieFlower",
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "marginRight": 8,
                    },
                  ]
                }
              >
                @designhumor
              </Text>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 20,
                      },
                      undefined,
                      Object {
                        "fontFamily": "FontAwesome",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
